Convert "Create a DevTools extension" to sample-based (#3616)
MicrosoftDocs/edge-developer: microsoft-edge/extensions/samples/custom-devtools-tool-code.md

---
title: Code for Custom DevTools tool
description: Explains the files and code in the Custom DevTools tool sample, a Microsoft Edge extension.  How the sample creates a custom tool tab and uses DevTools extension APIs.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: article
ms.service: microsoft-edge
ms.subservice: extensions
ms.date: 04/20/2026
---
# Code for Custom DevTools tool
<!-- https://learn.microsoft.com/microsoft-edge/extensions/samples/custom-devtools-tool-code -->
<!-- sibling 3 of 3 -->

<!-- todo: move this new file from /samples/ to /developer-guide/ -->

> The Custom DevTools Tool sample consists of the following files and code.  The code that interacts with the inspected webpage does the following:

Reviewer: ```suggestion
The Custom DevTools tool sample consists of the following files and code.  The code that interacts with the inspected webpage does the following:
```

* Displays memory information in the **Custom** DevTools tool.

* Displays the mouse-click position in the **Custom** DevTools tool, when the webpage is clicked.

* Displays a greeting alert in the inspected webpage, when the **Say Hello** button is clicked in the **Custom** DevTools tool.

See also:
* [Add a custom tool in Microsoft Edge DevTools](../developer-guide/devtools-extension.md)<!-- sibling 1 of 3 -->
   * [Sample: Custom DevTools tool](./custom-devtools-tool.md)<!-- sibling 2 of 3 -->

>       * [Code for Custom DevTools tool](./custom-devtools-tool-code.md)<!-- sibling 3 of 3 --> - the present article.

Reviewer: We're getting rid of the first article, as discussed. So this list of links can be simpler:

```suggestion
This article describes the code of the Custom DevTools tool sample. To first install and run the sample, see [Sample: Custom DevTools tool](./custom-devtools-tool.md)
```

```

The content script relays this info to the `devtools.js` file, where the data is displayed in DevTools in the **Custom** tool and in the **Console** tool.

<!-- ====================================================================== -->
## Overview of files
<!-- https://github.com/MicrosoftEdge/Demos/tree/main/devtools-extension/sample%204 -->

| File | Description |
|---|---|
| [manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/manifest.json) | Information about the extension: name, description, version, manifest version, and HTML page to show in DevTools. |
| [panel.html](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/panel.html) | Webpage to display in the custom panel in DevTools. |
| [devtools.html](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/devtools.html) | A non-rendered HTML file that's run when DevTools is opened, to load the extension's JavaScript files. |
| [service-worker.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/service-worker.js) | Service worker that sets up event listeners for communications between the inspected page and DevTools. |

> | [content_script.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/content_script.js) | Logic for the custom DevTools page.  Prints a message to the console when the script is injected in the page.  Adds a click event listener to the page that will send a message with mouse-click position in the inspected page. |

Reviewer: ```suggestion
| [content_script.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/content_script.js) | Code which the custom DevTools tool injects in the inspected webpage.  Prints a message to the console when the script is injected in the page.  Adds a click event listener to the page that will send a message with mouse-click position in the inspected page. |
```

> | [devtools.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/devtools.js) | Logic for the custom DevTools page. |

Reviewer: Let's move this up to be right below the devtools.html row in the table. These two files go together hand in hand.

| `permissions` | The local device requires permissions to view its system memory capacity, since the script calls API for that. |

`manifest.json`:

```json
{
    "name": "Custom DevTools Tool",
    "description": "A DevTools extension interacting with the inspected page",
    "manifest_version": 3,
    "version": "1.0",
    "devtools_page": "devtools.html",
    "content_scripts": [{
        "matches": [
            "http://*/*",
            "https://*/*"
        ],
        "run_at": "document_idle",
        "js": [
            "content_script.js"
        ]
    }],
    "background": {
        "service_worker": "service-worker.js"
    },
    "permissions": [
        "system.memory"
    ]
}
```

The `permissions` manifest member in `manifest.json` defines which permissions the extension requires from the user.  Some permissions are needed to use certain extension APIs.  The `system-memory` permission is required, in order to use the extension APIs.  Specific APIs have associated permissions.

> Keys within the "content_scripts" member:

Reviewer: ```suggestion
Keys within the `content_scripts` member:
```

    });
});
```

The above code connects the background service worker with the DevTools page.  It listens to when the DevTools page connects, saves the connection, and sends a response back to the DevTools page.

This is useful when your background service worker is collecting data or performing tasks in the background that you want to be available in your DevTools extension.

See also:
* [Content scripts](https://developer.chrome.com/docs/extensions/mv3/content_scripts/) in the Chrome Extensions docs.

<!-- ====================================================================== -->
## `content_script.js`

File in repo: 
* [content_script.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/content_script.js) - contains JavaScript that's injected into the inspected webpage (any webpage).  This file does the following:
   * Prints a message to the DevTools **Console** when the page is clicked.
   * Listens for the page click event via an event listener.
   * Adds a click event listener to the webpage; clicking the page sends an event, caught by the event listener, which then sends a message to the WebView2 Runtime.

> A _content script_ runs in the context of the inspected webpage.  In the same way that other scripts are loaded by the webpage, a content script has have access to the DOM and can change it.

Reviewer: ```suggestion
A _content script_ runs in the context of the inspected webpage.  In the same way that other scripts are loaded by the webpage, a content script has access to the DOM and can change it.
```

This code prints a message to the console when the script is injected in the page.  This code also adds a click event listener to the page that will send a message with mouse-click position in the inspected page by using the `chrome.runtime.sendMessage` API.

`content_script.js`:

```javascript
document.addEventListener("click", (event) => {
  chrome.runtime.sendMessage({
      click: true,
      xPosition: event.clientX + document.body.scrollLeft,
      yPosition: event.clientY + document.body.scrollTop
    },
    response => {
      console.log("Received response", response);
    }
  );
});
```

See also:
* [Content scripts](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts)

<!-- ====================================================================== -->

> ## `devtools.js`

Reviewer: Let's move this up to be after the devtools.html section. They go hand in hand.

When the user clicks anywhere in the inspected window, the DevTools extension will receive a message, from the background service worker, with `request.click == true` and the mouse position information.

See also: 
* [Supported APIs for Microsoft Edge extensions](../developer-guide/api-support.md)
   * [chrome.devtools.panels](https://developer.chrome.com/docs/extensions/reference/api/devtools/panels)
      * [create()](https://developer.chrome.com/docs/extensions/reference/api/devtools/panels#method-create)

<!-- ====================================================================== -->
## `icon.png`

[icon.png](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/icon.png) is the icon to display on the tool's tab in the **Activity bar** of DevTools and in the **More tools** menu:

![Icon file](./custom-devtools-tool-code-images/icon.png)

<!-- ====================================================================== -->
## `README.md`

> [README.md](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/README.md) contains basic information for developers about how to use the sample.

Reviewer: This file plays no role in the extension. It's just a nice-to-have for people visiting the github repo. I don't think we need this section here.
